chore(llvm): Build the whole LLVM stack together

EyeCantCU opened the pull request:
This PR refactors LLVM 19 and all of its components to build together within the same stack increasing maintainability

Other runtimes we previously built have been added as well (sans OpenMP for now due to issues with OpenSSF flags - we need to build Clang with those). This also introduces Clang's compiler runtime

Finally, libraries that are ABI compatible and share a consistent version between releases of LLVM are now built in the stack. Virtual provides have been added for compatibility, but will be progressively phased out

EyeCantCU committed:
chore(llvm): Build the whole LLVM stack together

This PR refactors LLVM 19 and all of its components to build together
within the same stack increasing maintainability

Other runtimes we previously built have been added as well (sans OpenMP
for now due to issues with OpenSSF flags - we need to build Clang with
those). This also introduces Clang's compiler runtime

Finally, libraries that are ABI compatible and share a consistent
version are now built in the stack. Virtual provides have been added for
compatibility, but will be progressively phased out

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

murraybd commented on llvm.yaml:
```diff
@@ -0,0 +1,830 @@
+      pipeline:
+        - uses: test/ldd-check
+          with:
+            packages: $(basename ${{targets.contextdir}})
```
You could use ${{subpkg.name}} here.

EyeCantCU replied:
Thank you, done!

jamonation commented:
@murraybd - is there anything preventing this from being merged from the OS side of things? 

e.g. I see rust 1.8x needs `libLLVM-19` at runtime, which this provides. Likewise clang-19, llvm-19 and llvm-19-dev.

murraybd commented on llvm.yaml:
```diff
@@ -0,0 +1,830 @@
+      provides:
+        - libclang-cpp-${{vars.major-version}}=${{package.full-version}}
+      runtime:
+        - libLLVM
```
We've seen some issues with abseil-cpp that leads us to believe depending on the specific version would be the safest thing e.g. `${{package.full-version}}` . See https://github.com/chainguard-dev/internal-dev/issues/9394#issuecomment-2688648434 for an example.

EyeCantCU replied:
Will do!

murraybd commented on llvm.yaml:
```diff
@@ -0,0 +1,830 @@
+      pipeline:
+        - runs: |
+            scan-build --help
+            scan-view --help
```
It looks like there are also binaries for `analyze-build` and `intercept-build` do those have arguments like `--help` or `--version` that we could test?

murraybd commented on llvm.yaml:
```diff
@@ -0,0 +1,830 @@
+      provides:
+        - py3-clang-${{vars.major-version}}=${{package.full-version}}
+      runtime:
+        - libLLVM
```
Again, I think specifying the version here would be safest.

EyeCantCU committed:
fix(llvm): Use explicit package versions

Also address the potential for the Python version used to generate
Clang's Python bindings drifting from the latest Python version

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
chore(llvm): Improve test coverage

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>

EyeCantCU committed:
fix(clang): Make libclang public

Signed-off-by: RJ Sampson <rj.sampson@chainguard.dev>
